inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/reduce_gpu_b_fs_yx_fsv16.cl

```diff
@@ -0,0 +1,429 @@
+    const uint out_idx = OUTPUT_GET_INDEX(b, f, x, y);
+#else
+    const uint out_idx = OUTPUT_GET_INDEX(b, f, y, x);
+#endif
+
+    const uint linear_idx = FUNC_CALL(calc_linear_offset)(b, f, y, x);
+    if (linear_idx >= COMPUTATIONAL_OPERATIONS_NUMBER)
+        return;
+
+    const uint input_x_pitch = FSV;
+    const uint input_y_pitch = input_x_pitch * (INPUT0_PAD_BEFORE_SIZE_X + INPUT0_SIZE_X + INPUT0_PAD_AFTER_SIZE_X);
+    const uint input_fs_pitch = input_y_pitch * (INPUT0_PAD_BEFORE_SIZE_Y + INPUT0_SIZE_Y + INPUT0_PAD_AFTER_SIZE_Y);    
+    const uint input_batch_pitch = input_fs_pitch * ((INPUT0_PAD_BEFORE_FEATURE_NUM + INPUT0_FEATURE_NUM + INPUT0_PAD_AFTER_FEATURE_NUM + FSV - 1) / FSV);
+
+    const uint output_y_pitch = FSV * (OUTPUT_PAD_BEFORE_SIZE_X + OUTPUT_SIZE_X + OUTPUT_PAD_AFTER_SIZE_X);
+
+#if REDUCE_BATCH
+    const uint batch_out = 0;
+    const uint batch_max_val = INPUT0_BATCH_NUM;
+#else
+    const uint batch_out = BATCH_NUM_IDX_COMP(linear_idx);
+    const uint batch_max_val = batch_out + 1;
+#endif
+
+#if REDUCE_FEATURE
+    const uint feature_out = 0;
+    const uint feature_max_val = INPUT0_FEATURE_NUM;
+#else
+    const uint feature_out = FEATURE_NUM_IDX_COMP(linear_idx);
+    const uint feature_max_val = feature_out + 1;
+#endif
+
+#if REDUCE_Y
+    const uint y_out = 0;
+    const uint y_max_val = INPUT0_SIZE_Y;
+#else
+    const uint y_out = SIZE_Y_IDX_COMP(linear_idx);
+    const uint y_max_val = y_out + 1;
+#endif
+
+#if REDUCE_X
+    const uint x_out = 0;
+    const uint x_max_val = INPUT0_SIZE_X / READ_OFFSET;
+    const uint x_leftover_start = x_max_val * READ_OFFSET;
+    const uint x_leftover_end = INPUT0_SIZE_X;
+#else
+    const uint x_out = SIZE_X_IDX_COMP(linear_idx);
+    const uint x_max_val = x_out + 1;
+    const uint x_leftover_start = x_out;
+    const uint x_leftover_end = x_max_val;
+#endif
+
+uint offset = batch_out * input_batch_pitch + ((feature_out + FSV - 1) / FSV) * input_fs_pitch + y_out * input_y_pitch + x_out * input_x_pitch;
+uint leftover_offset = batch_out * input_batch_pitch + ((feature_out + FSV - 1) / FSV) * input_fs_pitch + y_out * input_y_pitch + x_leftover_start * input_x_pitch;
+
+#if REDUCE_BFYX_BYX_BFX_FYX || REDUCE_BX_FX_YX_X
+    ACCUMULATOR_TYPE acc = INIT_VAL;
+    for (uint bi = batch_out; bi < batch_max_val; ++bi) {
+        for (uint fi = feature_out; fi < feature_max_val; fi += FSV) {
+            for (uint yi = y_out; yi < y_max_val; ++yi) {
+                for (uint xi = x_out; xi < x_max_val; ++xi) {
+                    INPUT_VEC input = (INPUT_VEC)(INPUT_INIT_VAL);
+                    #if (REDUCE_MAX_MODE || REDUCE_MIN_MODE || REDUCE_PROD_MODE || REDUCE_AND_MODE || REDUCE_LOG_SUM_EXP_MODE) && REDUCE_FEATURE && (INPUT0_FEATURE_NUM % FSV != 0)
+                        if (fi + get_sub_group_local_id() < INPUT0_FEATURE_NUM)
+                            for (int i = 0; i < READ_OFFSET; ++i)
```

Comment:
I don't think we want to loose all block reads just because there is leftover. If full sub-group is inside input x then we could still use block reads and only handle last feature slice with per work-item if.